### DISCUSSION TITLE
Quick way to build a Ghidra project with large number of files?

### DISCUSSION THREAD

**Discussion by sad-dev** [NationalSecurityAgency/ghidra#5530] [Q&A]
I have a large (thousands, totalling several GB) set of files that I would like to perform batch analysis on. What would be the fastest way to load all of them into a single project file and run autoanalysis on all of them? Can portions of this be parallelised?

**Comment by ryanmkurtz** [comment]
There's the GUI Batch Importer but that will just import them, not perform analysis.

You can use the Headless Analyzer with the `-import` flag which will by default perform an auto analysis.  I don't think you can parallelize that due to the project lock, but I could be wrong about that.

**Reply by dev747368** [reply]
Headless would also work for a shared project on a Ghidra server instance, which could have multiple instances connecting to it at the same time.  You will probably run into a scaling plateau at some point, but should be worth it.

**Reply by sad-dev** [reply]
So in short, to bypass the project lock, use GhidraServer with multiple instances of Ghidra to write multiple files at once?

I didn't see anything resembling "copy this program from this project to that project" so that may be (unfortunately) the best I've got...

**Reply by dev747368** [reply]
If you are thinking of doing bulk import/analysis with 1 binary per project, and then joining all the results into 1 mega-project at the end... that might be something worth while.

There is the ability to import a binary from a project into another.  In the GUI, you would just open your destination project like normal, and then in the front end window Project | View Project...., select the other project.  This will display the second project in a "READ-ONLY Project Data" window on the side that you can drag-n-drop or use the copy command to copy the binary to your main project.

I haven't dug into what code would be needed to duplicate that functionality, but I'm guessing its not rocket surgery.

**Comment by sad-dev** [comment]
> If you are thinking of doing bulk import/analysis with 1 binary per project, and then joining all the results into 1 mega-project at the end... that might be something worth while.
> 
> There is the ability to import a binary from a project into another. In the GUI, you would just open your destination project like normal, and then in the front end window Project | View Project...., select the other project. This will display the second project in a "READ-ONLY Project Data" window on the side that you can drag-n-drop or use the copy command to copy the binary to your main project.
> 
> I haven't dug into what code would be needed to duplicate that functionality, but I'm guessing its not rocket surgery.

You read my mind :laughing: I'll think on it and put up a feature/request for tools to streamline this workflow
